File: packages/attachment_engine/attachment_engine_macos/macos/attachment_engine_macos/Sources/attachment_engine_macos/PdfChannel.swift
// Copyright 2026 DHC Tech
// Use of this source code is governed by an MIT-style license that can be
// found in the LICENSE file.

import Cocoa
import FlutterMacOS
import PDFKit

/// Uses PDFKit (`PDFDocument`, `PDFPage.thumbnail`) to open a PDF and
/// render pages to PNG bitmaps, keyed by an opaque handle string so the
/// Dart side can hold multiple documents open concurrently. Identical
/// contract to the iOS implementation; only the bitmap-to-PNG encoding
/// step differs (`NSBitmapImageRep` instead of `UIImage.pngData()`, since
/// `PDFPage.thumbnail(of:for:)` returns `NSImage` on macOS).
///
/// Implements the Pigeon-generated `PdfHostApi` (see `Messages.g.swift`,
/// codegen'd from
/// `attachment_engine_platform_interface/pigeons/messages.dart`) instead
/// of a hand-written `FlutterMethodChannel`.
class PdfChannel: NSObject, PdfHostApi {
  private var openDocs: [String: PDFDocument] = [:]

  func register(with messenger: FlutterBinaryMessenger) {
    PdfHostApiSetup.setUp(binaryMessenger: messenger, api: self)
  }

  func unregister(with messenger: FlutterBinaryMessenger) {
    PdfHostApiSetup.setUp(binaryMessenger: messenger, api: nil)
    openDocs.removeAll()
  }

  func open(path: String) async throws -> PdfOpenResultMessage {
    guard let document = PDFDocument(url: URL(fileURLWithPath: path)) else {
      throw PigeonError(code: "open_failed", message: "Could not open PDF at \(path)", details: nil)
    }
    let handle = UUID().uuidString
    openDocs[handle] = document
    return PdfOpenResultMessage(handle: handle, pageCount: Int64(document.pageCount))
  }

  func renderPage(handle: String, index: Int64, width: Int64, height: Int64) async throws
    -> FlutterStandardTypedData
  {
    guard let document = openDocs[handle] else {
      throw PigeonError(code: "bad_handle", message: "Unknown PDF handle", details: nil)
    }
    guard let page = document.page(at: Int(index)) else {
      throw PigeonError(code: "render_failed", message: "Invalid page index \(index)", details: nil)
    }

    let pageRect = page.bounds(for: .mediaBox)
    let scale = min(
      Double(width) / Double(pageRect.width), Double(height) / Double(pageRect.height))
    let targetSize = CGSize(
      width: max(pageRect.width * CGFloat(scale), 1), height: max(pageRect.height * CGFloat(scale), 1))

    let image = page.thumbnail(of: targetSize, for: .mediaBox)
    guard let data = Self.pngData(from: image) else {
      throw PigeonError(code: "render_failed", message: "PNG encoding failed", details: nil)
    }
    return FlutterStandardTypedData(bytes: data)
  }

  func close(handle: String) async throws {
    openDocs.removeValue(forKey: handle)
  }

  /// `NSImage` has no `pngData()` equivalent to `UIImage`; go through its
  /// best (largest) bitmap representation and `NSBitmapImageRep`'s PNG
  /// encoder instead.
  private static func pngData(from image: NSImage) -> Data? {
    guard
      let tiffData = image.tiffRepresentation,
      let bitmap = NSBitmapImageRep(data: tiffData)
    else {
      return nil
    }
    return bitmap.representation(using: .png, properties: [:])
  }
}

File: packages/attachment_engine/attachment_engine_android/android/src/main/kotlin/com/dhc/attachment_engine/PdfChannel.kt
package com.dhc.attachment_engine

import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import io.flutter.plugin.common.BinaryMessenger
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Replaces `pdfx`. Uses [android.graphics.pdf.PdfRenderer] to open a PDF
 * and render pages to PNG bitmaps, keyed by an opaque handle string so the
 * Dart side can hold multiple documents open concurrently.
 *
 * Implements the Pigeon-generated [PdfHostApi] (see the codegen'd
 * `Messages.g.kt`, produced from
 * `attachment_engine_platform_interface/pigeons/messages.dart`) instead of
 * a hand-written `MethodChannel`.
 */
class PdfChannel : PdfHostApi {
    private data class OpenDoc(val fd: ParcelFileDescriptor, val renderer: PdfRenderer)

    private val openDocs = mutableMapOf<String, OpenDoc>()

    fun register(messenger: BinaryMessenger) {
        PdfHostApi.setUp(messenger, this)
    }

    fun unregister(messenger: BinaryMessenger) {
        PdfHostApi.setUp(messenger, null)
        openDocs.values.forEach {
            try {
                it.renderer.close()
                it.fd.close()
            } catch (_: Exception) {
            }
        }
        openDocs.clear()
    }

    override suspend fun open(path: String): PdfOpenResultMessage =
        withContext(Dispatchers.IO) {
            val file = File(path)
            val fd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
            val renderer = PdfRenderer(fd)
            val handle = UUID.randomUUID().toString()
            openDocs[handle] = OpenDoc(fd, renderer)
            PdfOpenResultMessage(handle = handle, pageCount = renderer.pageCount.toLong())
        }

    override suspend fun renderPage(
        handle: String,
        index: Long,
        width: Long,
        height: Long,
    ): ByteArray =
        withContext(Dispatchers.IO) {
            val doc = openDocs[handle] ?: throw FlutterError("bad_handle", "Unknown PDF handle", null)
            val safeWidth = width.toInt().coerceAtLeast(1)
            val safeHeight = height.toInt().coerceAtLeast(1)
            val page = doc.renderer.openPage(index.toInt())
            try {
                val scale = minOf(safeWidth.toDouble() / page.width, safeHeight.toDouble() / page.height)
                val bmpWidth = (page.width * scale).toInt().coerceAtLeast(1)
                val bmpHeight = (page.height * scale).toInt().coerceAtLeast(1)
                val bitmap = Bitmap.createBitmap(bmpWidth, bmpHeight, Bitmap.Config.ARGB_8888)
                bitmap.eraseColor(Color.WHITE)
                page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
                val stream = ByteArrayOutputStream()
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
                stream.toByteArray()
            } finally {
                page.close()
            }
        }

    override suspend fun close(handle: String) {
        withContext(Dispatchers.IO) {
            val doc = openDocs.remove(handle)
            try {
                doc?.renderer?.close()
                doc?.fd?.close()
            } catch (_: Exception) {
            }
        }
    }
}
